Add Kubernetes access section to the role editor

bl-nero opened the pull request:
**To turn on the new UI, go to developer tools -> Application -> Local storage and add a `grv_teleport_use_new_role_editor` key set to `true`. This will be lifted once UI is good to be released.**

[Figma mocks](https://www.figma.com/design/v6GunK50D2VC7w7I2FBDNf/Access-(Management)?node-id=5021-39478&m=dev)

![Screenshot 2024-10-17 at 19 47 06](https://github.com/user-attachments/assets/39c99c6c-4955-4a0a-8343-821e97d76338)

Deviations from the design:
- We don't have the allow/deny switch just yet
- Used `H4` instead of `H3` for the resources header
- The "add a resource" button is slightly bigger, since it's consistent with the labels picker. I'll consider modifying them both separately.
- Not all resource fields are combo boxes.
- Kind is not a multi-select field.
- No numbering (or, should I say, "lettering"?) in resource field names

bl-nero committed:
Add Kubernetes access section to the role editor

ravicious commented on web/packages/teleport/src/Roles/RoleEditor/StandardEditor.tsx:
```diff
@@ -419,12 +437,166 @@ export function ServerAccessSpecSection({
+      />
+
+      <Flex flexDirection="column" gap={3} mt={3}>
+        {value.resources.map((resource, iRes) => (
```
Nit, but why `iRes` and not `index` or `resourceIndex`? When I saw `resources.map((resource, iRes) => …`, I assumed that `resources` is a map because I didn't quite understand at first what `iRes` is.

ravicious commented on web/packages/teleport/src/Roles/RoleEditor/StandardEditor.tsx:
```diff
@@ -419,12 +437,166 @@ export function ServerAccessSpecSection({
+            onRemove={() =>
+              onChange?.({
+                ...value,
+                resources: value.resources.toSpliced(iRes, 1),
```
Oh nice, `toSpliced` is pretty cool, it might be the first time I see it. It caniuse.com says that [Chrome for Android got support for it just in September this year](https://caniuse.com/mdn-javascript_builtins_array_tospliced), but I guess it should be safe? I imagine that most of the time Chrome is updated separately from the OS.

I looked up `webassets/teleport/app/app.js` and it doesn't look like this gets compiled to a polyfilled version, it's just raw in there.

bl-nero replied:
@ravicious No, for some reason, caniuse.com only shows the last Android Chrome version (it's even [trickling down as a bug to other projects](https://github.com/browserslist/browserslist/issues/156#issuecomment-315416975)). It's been there since version 110, which was released in February 2023 (see [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/toSpliced#browser_compatibility)).

That being said, if mobile browsers are on our compatibility list, it should be documented [in the readme](https://github.com/gravitational/teleport/blob/eda3671595b6507c695dd460ae9d659f0505654a/web/README.md?plain=1#L177). Let me know if you have a more complete list of targets.

ravicious replied:
That's good to hear regarding caniuse.com. I've just linked to it under another PR and was surprised to see Android Chrome again. Feb 2023 is good enough.

Regarding the target list, I'd like us to change it but I haven't had the time to advocate for that. It was set, in my opinion, [rather hastily](https://gravitational.slack.com/archives/C02DQ1C2BMW/p1657730403674659) (here's the PR it's been part of https://github.com/gravitational/webapps/pull/977). If you [plug it into `browsersl.ist`](https://browsersl.ist/#q=last+2+chrome+version%2C+last+2+edge+version%2C+last+2+firefox+version%2C+last+2+safari+version), you'll see that at the moment we "support" only 3.7% of global audience. Even if you filter that down to only North America, that's still just 4.8%. I think that's way too low, compared to 87.1% coverage that using the `defaults` target brings.

ravicious commented on web/packages/teleport/src/Roles/RoleEditor/StandardEditor.tsx:
```diff
@@ -419,12 +437,166 @@ export function ServerAccessSpecSection({
+      <Flex flexDirection="column" gap={3} mt={3}>
+        {value.resources.map((resource, iRes) => (
+          <KubernetesResourceView
+            key={iRes}
```
[Using indexes as keys is not recommended](https://react.dev/learn/rendering-lists#why-does-react-need-keys), this is particularly true for this form as you can remove an item at any position from the model.

Would it be possible to add a `uuid` field to `KubernetesResourceModel`, generate it in `newKubernetesResourceModel` and then make sure it doesn't make its way to the backend? I haven't looked at the code too closely yet, but I assume at some point we need to map values from the UI to some kind of yaml.

ravicious replied:
I suppose the uuid field could be just ignored within `roleEditorModelToRole`?

bl-nero replied:
Oh, crap. I'm well aware that this is an anti-pattern (as a matter of fact, I was commenting on the similar thing in another PR), but somehow still fell into this trap here. :) Yes, I'll add an ID to the editor model.

ravicious commented on web/packages/teleport/src/Roles/RoleEditor/StandardEditor.tsx:
```diff
@@ -419,12 +437,166 @@ export function ServerAccessSpecSection({
   pointer-events: ${p => (p.mute ? 'none' : '')};
 `;
+
+// TODO(bl-nero): This should ideally use tonal neutral 1 from the opposite
+// theme as background.
```
Does the BBLP theme have an opposite theme? I wonder if each theme should have the concept of an inverse colors in itself.

bl-nero replied:
@ravicious We decided with Kenny that I'll later cherry-pick tonal neutral colors from each theme the inverse theme, treating light as inverse of BBLP. (Cherry-picking is the only option, since we can't just switch the color theme entirely for a subtree of components — exactly for the reason you described.) This should be a good solution until Kenny provides some more concrete guidelines on the `Mark` component itself.

ravicious reviewed (approved):
It's the first time I'm looking at the code behind the role editor and I have to say I really like how it's written, it's pretty clean!

Left a couple of minor suggestions, with the most "major" being the one about not using indexes as keys.

bl-nero commented on api/types/constants.go:
```diff
@@ -1247,7 +1247,11 @@ var RequestableResourceKinds = []string{
+// is for the latest version of Role resources; roles whose version is set to
+// v6 or prior only support [KindKubePod].
+// This list needs to be kept in sync with `kubernetesResourceKindOptions` in
+// `web/packages/teleport/src/Roles/RoleEditor/standardmodel.ts`.
```
I simply separated this to a separate comment above the godoc one.

bl-nero commented on web/packages/teleport/src/Roles/RoleEditor/StandardEditor.tsx:
```diff
@@ -419,12 +437,166 @@ export function ServerAccessSpecSection({
+          >
+            <Icon.Add disabled={isProcessing} size="small" />
+            {value.resources.length > 0
+              ? 'Add Another Resource'
+              : 'Add a Resource'}
```
Yes, you're correct that the labels input is "at fault" here. The thing is, we historically had these inconsistencies pretty much everywhere, so I decided to cut it short when pushing through the change where we got rid of all-caps buttons, and documented it as a [practice to be followed](https://github.com/gravitational/teleport/blob/master/web/packages/design/README.md#other-advice). At least Kenny is on board with this rule, so I guess that these inconsistencies in the design are purely incidental.

bl-nero committed:
Review

bl-nero committed:
Merge branch 'master' into bl-nero/role-editor-4

kiosion commented on web/packages/teleport/src/Roles/RoleEditor/standardmodel.ts:
```diff
@@ -211,25 +409,52 @@ export function roleEditorModelToRole(roleModel: RoleEditorModel): Role {
       }
-      role.spec.allow.node_labels = labels;
-      role.spec.allow.logins = spec.logins.map(opt => opt.value);
+      default:
+        kind satisfies never;
```
maybe I'm overlooking something, but does this have any effect?

bl-nero replied:
@kiosion Yes, that's a very useful pattern that I learned a couple of months ago: `kind satisfies never` means that at this point, the `kind` variable cannot possibly hold any value. Using it in a `default` block of a `case` statement means that (at least from the TS Compiler's perspective) every possible case has been handled. Thus, adding another possibility to the union type of `kind` would cause a compiler error and enforce handling this case before the code can be merged.

bl-nero replied:
(That's one of the things that I've been missing from languages like Rust, and I'm very happy that there's a way to express it in TS.)

kiosion reviewed (approved):
lgtm!

bl-nero committed:
Merge branch 'master' into bl-nero/role-editor-4

bl-nero committed:
Update the k8s operator docs

This also removes the mention of valid values of the Kind field, as I don't
want the external documentation to point to Teleport source files.

bl-nero committed:
Update operator CRDs and Terraform resources
